### ISSUE TITLE
[HOLD for payment 2023-12-15] MEDIUM: [$500] Notifications - Reports removed from LHN can still show notifications leading to unremovable notifications

### ISSUE THREAD

**Issue by kbecciv** [Expensify/App#28536]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:

**For Web and Desktop:**
1. Send a message in any report
2. Hover over the message and reply in thread
3. Right click on the thread and click mark as unread
4. Click out of the thread and notice a notification that you can no longer remove
5. Additionally if you remove the original message there's is no longer any possible way of removing the notification

**For mobile**
1. Go to any report and hold press on any message and reply in thread
2. While in the thread, click the three dots and click "pin"
3. Go back to the LHN list of reports and hold press the empty pinned thread and choose mark as unread
4. Hold press again on the pinned thread and choose unpin
5. Notice the thread getting hidden but the notification remaining

## Expected Result:
The user should not be stuck with notifications that are impossible to remove.

## Actual Result:
User is stuck with notifications that are impossible to remove, even unintentionally, I have notifications that I can't remove on all of my accounts.

## Workaround:
Unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android / native
- [x] Android / Chrome
- [ ] iOS / native
- [x] iOS / Safari
- [x] MacOS / Chrome / Safari
- [x] MacOS / Desktop

**Version Number:** 1.3.75.2
**Reproducible in staging?:** y
**Reproducible in production?:** y
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** Any additional supporting documentation

https://github.com/Expensify/App/assets/93399543/a560de5f-9ec4-4183-bdf0-05a04aa79dfc

https://github.com/Expensify/App/assets/93399543/bfb77f4e-2574-48eb-bbd9-05f89fbe56fe

https://github.com/Expensify/App/assets/93399543/e3be3eb4-7388-49de-877b-dd2a1412a3df

https://github.com/Expensify/App/assets/93399543/93897821-13cb-4b31-b71c-0e6208e4c4d2

https://github.com/Expensify/App/assets/93399543/97b614e0-a00c-4246-a705-4a35cf1de9c9

**Expensify/Expensify Issue URL:**
**Issue reported by:** @AmjedNazzal 
**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1695916782976289

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~01082d000584ab5766</li>
        <li>Upwork Job ID: 1708128118997794816</li>
        <li>Last Price Increase: 2023-09-30</li>
<li>Automatic offers: </li>
<ul>
<li>cubuspl42 | Reviewer | 27217153</li>
<li>AmjedNazzal | Reporter | 27217154</li>
</ul></li>
    </ul>
</details>

**Comment by kbecciv** [comment]
## Proposal
[AmjedNazzal](https://github.com/AmjedNazzal) on GH
### Please re-state the problem that we are trying to solve in this issue.
Reports removed from LHN can still show notifications leading to unremovable notifications
### What is the root cause of that problem?
This is a problem with the logic of the notifications system colliding with the logic of LHN, in LHN we remove any report/thread that does not have replies in it by using isEmptyChat in ReportUtils.shouldReportBeInOptionList, on the other hand, ReportUtils.isUnread function takes all of the reports, doesn't check if reports are hidden or not and just shows notifications for any report where lastReadTime < lastVisibleActionCreated leading to this issue.
### What changes do you think we should make in order to solve the problem?
ReportUtils.isUnread should not show notifications for hidden reports or reports with no replies since that doesn't make sense and since the user can no longer click on them or mark them as read.
```javascript
function isUnread(report) {
..
const lastVisibleMessage = ReportActionsUtils.getLastVisibleMessage(report.reportID);
const isEmptyChat = !report.lastMessageText && !report.lastMessageTranslationKey && !lastVisibleMessage.lastMessageText && !lastVisibleMessage.lastMessageTranslationKey;
if (isEmptyChat) {
    return false;
}
```
To keep things DRY we can move const lastVisibleMessage and const isEmptyChat to a higher level so we can use isEmptyChat in both ReportUtils.isUnread and ReportUtils.shouldReportBeInOptionList
### Result
https://github.com/Expensify/App/assets/74202751/036da4e1-e0c1-412b-8561-98e16ab4a73c

**Comment by AmjedNazzal** [comment]
@allroundexperts What's the point of letting the user mark something as unread when the user is A) Viewing that thread already, and B) That thread is empty? if anything we should just remove mark as unread for empty threads or at least ignore the marking as unread to begin with because otherwise there will be a whole bunch of hidden reports that are marked as unread in the Onyx store

**Comment by cubuspl42** [comment]
@AmjedNazzal @allroundexperts 

What's the use of those reports that are hidden in the LHN? Does the app reason about them at all? In which places we want _not_ to exclude them from any processing?

**Comment by AmjedNazzal** [comment]
@cubuspl42 I don't know but what we seem to be doing is just hide empty threads rather than removing them from the store: 

```const lastVisibleMessage = ReportActionsUtils.getLastVisibleMessage(report.reportID);
    const isEmptyChat = !report.lastMessageText && !report.lastMessageTranslationKey && !lastVisibleMessage.lastMessageText && !lastVisibleMessage.lastMessageTranslationKey;
    const canHideReport = shouldHideReport(report, currentReportId);

    // Hide only chat threads that haven't been commented on (other threads are actionable)
    if (isChatThread(report) && canHideReport && isEmptyChat) {
        return false;
    }```

**Comment by AmjedNazzal** [comment]
@cubuspl42 If removing empty threads is a better solution, what we can do is create a utility to check if there are threads that should be removed:
 
```javascript
function shouldRemoveThread(report, currentReportId) {
    const lastVisibleMessage = ReportActionsUtils.getLastVisibleMessage(report.reportID);
    const isEmptyChat = !report.lastMessageText && !report.lastMessageTranslationKey && !lastVisibleMessage.lastMessageText && !lastVisibleMessage.lastMessageTranslationKey;
    const isCurentReport = report.reportID === currentReportId;

    if (isChatThread(report) && isEmptyChat && !isCurentReport) {
        return true;
    }  
}
```

We can call this in SidebarLinksData whenever a change happens in the LHN meaning an empty thread could've been created and then hidden, and from there we could remove the empty thread from the onyx: 

```javascript
useEffect(() => {
        
        if (currentReportID) {
        const threadsToRemove = _.filter(chatReports, (report) => ReportUtils.shouldRemoveThread(report, currentReportID));
        
        threadsToRemove.forEach((thread) => {
            const { reportID } = thread;
            
            Report.leaveRoom(reportID);
          });
        }
      }, [optionListItemsWithCurrentReport]);
```

**Comment by AmjedNazzal** [comment]
@cubuspl42 Whispers is something like: 
![Moderation Flagging Options](https://help.expensify.com/assets/images/moderation-flag-page.png){:width="100%"}

Depending on the severity of the offense, messages can be hidden (with an option to reveal) or fully removed, and in extreme cases, the sender of the message can be temporarily or permanently blocked from posting.

You will receive a whisper from Concierge any time your content has been flagged, as well as when you have successfully flagged a piece of content.

It's messages that only you can see that are sent to you automatically I think

**Comment by AmjedNazzal** [comment]
I don't think whispers are relevant here, we are hiding empty threads even when there's a whisper in them.

**Comment by tjferriss** [comment]
Coming in late, but I am able to reproduce the bug.

**Comment by cubuspl42** [comment]
The existing proposals are still not satisfying. In my opinion, we should introduce a notion of _active_ reports and:

- Only present the active reports on the LHN
- Only consider the active reports when displaying the unread indicator
- ...and possibly only consider active reports in other places (?)

In one of the previous comments, I used the adjective "existing", but "active" may be a better name. Still, it's more about a concept. And pragmatically, it would end up as a helper in `ReportUtils`.

I'm not suggesting something with significant observable differences from what was proposed so far, but rather I disagree with the way things were _modeled_ in the solutions proposed so far.

Please let me know if you see any flaw in this model.

**Comment by AmjedNazzal** [comment]
@cubuspl42 we can consolidate both calls into one utility like so: 

```javascript
function isActiveAndUnread(report, currentReportID = null) {
    return shouldReportBeInOptionList(report, currentReportID) && isUnread(report);
}
```
 And from there we can call this whenever needed, for example: 
 
 ```javascript
Onyx.connect({
    key: ONYXKEYS.COLLECTION.REPORT,
    waitForCollectionCallback: true,
    callback: (reportsFromOnyx) => {
        const unreadReports = _.filter(reportsFromOnyx, report => ReportUtils.isActiveAndUnread(report));
        updateUnread(_.size(unreadReports));
    },
});
```

```javascript
useEffect(() => {
        if (currentReportID) {
            const unreadReports = _.filter(chatReports, report => ReportUtils.isActiveAndUnread(report, currentReportID));
            updateUnread(_.size(unreadReports));
        }
}, [optionListItemsWithCurrentReport]);
```

To clarify, the reason I'm calling this utility in SidebarLinksData is to insure that any changes in the list of reports such as a report being added or the currentReportID changing is captured and used to determine the shown notifications.

**Comment by cubuspl42** [comment]
Okay, that's not exactly what I meant, but I'm starting to feel that I might have focused on details that can be a part of PR-level discussion.

The fundamental idea of the proposal by @AmjedNazzal makes sense to me.

Thank you, @allroundexperts, for testing it and finding the issue you described! This will need to be handled. At the same time, the primary solution proposed by you (`ReportUtils.isUread(report) && ReportUtils.isEmptyReport(report)`) is a slightly less elegant solution in my opinion (as the logic is not shared), while the alternative one (`ReportUtils.shouldReportBeInOptionList(report, null, true, ..., true)`) is quite similar to the original proposal by @AmjedNazzal.

Thank you for both proposals and the following discussion!

I say we go with @AmjedNazzal

C+ reviewed 🎀 👀 🎀

**Comment by AmjedNazzal** [comment]
@cubuspl42 Actually to be fair I updated my proposal at some point and I happened to also use `const unreadReports = _.filter(reportsFromOnyx, (report) => ReportUtils.shouldReportBeInOptionList` but @allroundexperts didn't change his proposal I think so if that's the main focus then he used it first, my edited proposal was less about the specific use of const unreadReports = _.filter(reportsFromOnyx, (report) => ReportUtils.shouldReportBeInOptionList and more about updating the unread in SidebarLinksData too as I felt it makes sure everything is working in a consistent way between the LHN current view and the unread status. if the main idea here is the specific usage of ReportUtils.shouldReportBeInOptionList in unreadReports within UnreadIndicatorUpdater then I wasn't the first to do that. again my updated proposal was more concerned with how to tie in LHN current view and the unread count based on the most current changes in the LHN.

**Comment by cubuspl42** [comment]
I didn't notice that the first proposal (by @AmjedNazzal) was at some point inspired by the later one (by @allroundexperts).

In my opinion, it is crucial to share as much logic as possible between the presentation of the reports in the LHN and the update indicator, as in the user mental model (as I see it), they relate _to the same thing_. I'm notified about some updates to the things _I see_. If I'm notified about something _I can't see_, this is a paradox and leads to issues like this one.

As it's now clear that the first contributor to propose such solution was @allroundexperts, I'd like to change my recommendation to @allroundexperts.

C+ reviewed 🎀 👀 🎀 (again)

Thanks, everyone, for the honesty, but let's use the "## Proposal [Updated]" marker next time when one proposal takes strong inspiration from the other.

**Comment by cubuspl42** [comment]
@Gonals Bump

**Comment by Gonals** [comment]
Assigned!

**Comment by AmjedNazzal** [comment]
Bump, any updates?

**Comment by cubuspl42** [comment]
Oh my 😳

**Comment by cubuspl42** [comment]
@AmjedNazzal Thanks! I bumped a question on the PR. I'm not sure why Melvin gave up on this issue 😶

**Comment by AmjedNazzal** [comment]
@cubuspl42 No worries! yeah I noticed Melv has been sometimes letting go of things especially when they become a little old 😅

**Comment by iwiznia** [comment]
FWIW I am using staging web and the bug is still there (same happens in latest mode):
<img width="383" alt="image" src="https://github.com/Expensify/App/assets/521248/602625b4-2f06-40f9-92d9-3ac592e2c8ca">

**Comment by tjferriss** [comment]
Waiting on hold period before sending payment

**Comment by cubuspl42** [comment]
- The PR that introduced the bug has been identified. Link to the PR:
    - This can be treated as a clarification of the expected behavior; the notifications referred to existing entities that just couldn't be navigated to
- The offending PR has been commented on, pointing out the bug it caused and why, so the author and reviewers can learn from the mistake. Link to comment:
    - N/A
- A discussion in #expensify-bugs has been started about whether any other steps should be taken (e.g. updating the PR review checklist) in order to catch this type of bug sooner. Link to discussion:
    - No need for additional discussion
- Determine if we should create a regression test for this bug.
    - Up to the QA team
- If we decide to create a regression test for the bug, please propose the regression test steps to ensure the same bug will not reach production again.
    - Note your notification count (let's say it's 10)
    - Send a message in any report
    - Replay in a thread to that message
    - Mark the first message in the thread (the one you're replying to) as unread
        - [ ] Verify that the notification count increased (in our case, to 11)
    - Navigate out of the thread
        - [ ] Verify that you can't see the thread just left in the LHN
        - [ ] Verify that the notification count is now as initially (in our case, 10)

**Comment by tjferriss** [comment]
@cubuspl42 @AmjedNazzal payments have been made via Upworks

**Comment by allroundexperts** [comment]
Hi @tjferriss!

This got merged after about a month. Back then, we had a late merge penalty. So this is not eligible for any payment to me and @cubuspl42. 

However, I think the delay was on my part and @cubuspl42 was pretty prompt throughout. I would request to not apply the late penalty for him. It's fine if it applies to me.

**Comment by tjferriss** [comment]
>However, I think the delay was on my part and @cubuspl42 was pretty prompt throughout. I would request to not apply the late penalty for him.

I'll agree with you. I think we're all set here.

**Comment by cubuspl42** [comment]
@allroundexperts Thank you for these words.

From my perspective, this is a workflow thing. My primary tool for handling all issues I'm assigned to is the GitHub inbox. When some action is needed on the Contributor side, I inform them or ask a question and mark the notification as "Done".

Very rarely (it might be the first or second time), the Contributor misses the question and I move on. The issue and PR get forgotten.

I'm not sure if it's Melvin's malfunction or just a blind spot in Melvin's logic. It would be great if he could remind us about such stalled issues.

**Comment by allroundexperts** [comment]
That does happen but in this case, I think it was more of my fault. I saw your reminder but was at an airport (preparing for a 5h flight) and just completely forgot about this 😄

**Comment by cubuspl42** [comment]
@tjferriss Hm, as I understand this...

- either you _disagree_ with @allroundexperts and we're all set here
- ...or you _agree_ with him and we hold on the C+ payment (mine)

Unless I missed something

**Comment by cubuspl42** [comment]
Taking into consideration that this [caused a regression](https://github.com/Expensify/App/pull/30133/files#r1434388712), and the time it took to merge, let's just agree we're all set and payments do not apply.

Let' move on.
